ahorro_combustible/utils.py

```diff
@@ -0,0 +1,61 @@
+from typing import Dict, Tuple, Sequence
+
+
+def count_brands(gas_stations) -> Sequence[Tuple[int, str]]:
+    '''Count all the brands in the `gas_stations` input. It can be considered that
+    the ``Rotulo`` entry contains the brand name of the gas station.
+
+    Parameters
+    ----------
+    gas_stations : dict
+        A dictionary that contains at least the equivalent structure to the
+        following example.
+
+        { 'ListaEESSPrecio': [ { 'Rótulo': 'ES DULANTZI REPSOL' }, ... ]}
+
+    Returns
+    -------
+    Sequence[Tuple[int, str]]
+        A sorted sequence of tuples being the first part of each tuple the count
+        number of the second part, the brand name.
+
+        The items will be sorted by the count number (highest, first) and the
+        brand name.
+
+    Examples
+    --------
+
+    >>> count_brands({'ListaEESSPrecio': [ {'Rótulo': 'A' },
+    ...                                    {'Rótulo': 'B' },
+    ...                                    {'Rótulo': 'A' }]})
+    [(2, 'A'), (1, 'B')]
+
+    >>> count_brands({'ListaEESSPrecio': [ {'Rótulo': 'A' },
+    ...                                    {'Rótulo': 'C' },
+    ...                                    {'Rótulo': 'C' },
+    ...                                    {'Rótulo': 'B' }]})
+    [(2, 'C'), (1, 'B'), (1, 'A')]
+    '''
+    # An empty list to add tuples with brands and petrol stations quantity
+    
+    select_petrol_stations = []
+    
+    # Through this loop I take the brands and petrol station quantity
+    
+    for petrol_station in gas_stations['ListaEESSPrecio']:
+        if "Rótulo" in petrol_station:
+            select_petrol_stations.append(petrol_station["Rótulo"])
+            
+    # Another empty list to add the final result without selection and order
+          
+    lst_tup_sorted = []
+    
+    # Next loop: to number, remove duplicates and sort
+    
+    for petrol_st in select_petrol_stations:
```

Comment:
This loop is quite inefficient. As shown below execution time is exponential in regards to the size of the gas station list. A better solution is possible!

Execution time for size 100 was 0.000407564002671279
Execution time for size 1000 was 0.022546574007719755
Execution time for size 10000 was 1.8891225090046646
Execution time for size 100000 was 162.07698707901

Reply:
I have done a few changes. I have got to reduce 10 seconds from the original time of 27 seconds, so the code takes 17 seconds to extract the sorted tuples list.